### PR TITLE
PERF: lib.generate_slices

### PR DESCRIPTION
**Pull request by mzeitlin11**
```
import numpy as np
import pandas._libs.lib as lib

np.random.seed(0)
inds = np.random.randint(0, 100, 1000000)
ngroups = inds.max() + 1

%timeit lib.generate_slices(inds, ngroups)
```
Master:
`1.86 ms ± 44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)`

This pr:
`859 µs ± 127 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)`

Adding the `cython` decorators makes a huge difference here

### CODE REVIEW

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Comment by jreback** [comment]
thanks @mzeitlin11

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -920,19 +921,20 @@ def generate_slices(const intp_t[:] labels, Py_ssize_t ngroups):
...
+                start += 1
+            else:
+                group_size += 1
+                if i == n - 1 or lab != labels[i + 1]:
+                    starts[lab] = start
+                    ends[lab] = start + group_size
+                    start += group_size
+                    group_size = 0
+
+    return np.asarray(starts), np.asarray(ends)
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:937]: FYI there is a tiny perf bump from using `starts.base` instead of `np.asarray(starts)` in these cases.

Comment (mzeitlin11) [reply pandas/_libs/lib.pyx:937]: Ah thanks, good to know

**Pull request merged**